lib/web_catalog/api_extractor.es6.js

```diff
@@ -0,0 +1,406 @@
+        for (let i = 0; i < objectProp.length; i++) {
+          this.buildInObjectProperties.push(this.cleanUp_(objectProp[i]));
+        }
+      },
+    },
+    {
+      name: 'getClassName_',
+      description: 'Search object id\'s prototype chain to find the the' +
+        'class it belongs to. The class is the prototype with a' +
+        'constructor property',
+      code: function(id, og) {
+        if (id === null || og.getType(id) !== 'object') return null;
+        let ctorId = og.lookup('+constructor+', id);
+        if (ctorId) {
+          return og.getFunctionName(ctorId);
+        }
+        return this.getClassName_(og.getPrototype(id), og);
+      },
+    },
+    {
+      name: 'postProcess_',
+      description: 'Remove blacklist properties and add' +
+        'Function and Object\'s APIs.',
+      code: function(apiCatalogs, og) {
+        for (let i = 0; i < this.blacklistProperties.length; i++) {
+          let prop = this.blacklistProperties[i];
+          if (apiCatalogs[prop]) {
+            delete apiCatalogs[prop];
+          }
+        }
+        // Add function and object blacklisted property
+        // to Function and Object interface.
+        if (!apiCatalogs[og.getFunctionName(this.metadata.functionId)]) {
+          apiCatalogs[og.getFunctionName(this.metadata.functionId)] = [];
+        }
+        apiCatalogs[og.getFunctionName(this.metadata.functionId)] =
+          apiCatalogs[og.getFunctionName(this.metadata.functionId)]
+            .concat(this.buildInFunctionProperties);
+        if (!apiCatalogs[og.getFunctionName(this.metadata.objectId)]) {
+          apiCatalogs[og.getFunctionName(this.metadata.objectId)] = [];
+        }
+        apiCatalogs[og.getFunctionName(this.metadata.objectId)] =
+          apiCatalogs[og.getFunctionName(this.metadata.objectId)]
+            .concat(this.buildInObjectProperties);
+      },
+    },
+    {
+      name: 'extractAPI_',
+      description: 'extractAPI extract API for interface with a given id.',
+      /*
+        @param {ObjectGraph} og - The object graph.
+        @param {JSON} map - the json to store web api for interface
+        @param {int} id - the id currently visiting.
+        @param {String} name - the property name from its parent node.
+        @param {JSON} visited - a map of visted properties.
+        @param {JSON} opt - store additional infomation about this object.
+            opt.firstLevel: if the object is a first level,
+                first level objects are exists under window object.
+                False if undefined.
+            opt.proto: if this object is a __proto__ of other object.
+                False if undefined.
+
+      */
+      code: function(og, map, id, name, visited, opt) {
+        opt = opt || {};
+        // First level means this object exists as window.X.
+        let firstLevel = opt.firstLevel || false;
+        // Proto means this object is a other object's __proto__
```

Comment:
The only situation I noticed that the object.__proto__ are visited before object is when object.__proto__ is also a first level object, such as AudioContext, BaseAudioContext and EventTarget. Since they are first level object, the visit order does not change the result.
 Is it possible that some obejct's __proto__ is another object's property?